README.md
# md2html
**Python Markdown to single-file HTML convertor**

```
usage: md2html.py [-h] [-C CSS_FILE [CSS_FILE ...]] [-W BASE_FOLDER]
                  [-T HTML_TEMPLATE] [-F FAVICON] [-B BACKGROUND_IMG]
                  [-L LOGO] [-M MARKDOWN_EXTENSION [MARKDOWN_EXTENSION ...]]
                  [-O OUTPUT_FILE]
                  input_file

positional arguments:
  input_file            Markdown input file

optional arguments:
  -h, --help            show this help message and exit
  -C CSS_FILE [CSS_FILE ...], --css CSS_FILE [CSS_FILE ...]
                        CSS files
  -W BASE_FOLDER, --working-dir BASE_FOLDER
                        Working base directory where /css and /layout folders are found
  -T HTML_TEMPLATE, --template HTML_TEMPLATE
                        HTML template for Jinja2, defaults to template.html
  -F FAVICON, --favicon FAVICON
                        PNG/BMP/JPG favicon
  -B BACKGROUND_IMG, --background BACKGROUND_IMG
                        PNG/BMP/JPG background image
  -L LOGO, --logo LOGO  Logo image (PNG/BMP/JPG)
  -M MARKDOWN_EXTENSION [MARKDOWN_EXTENSION ...], --extensions MARKDOWN_EXTENSION [MARKDOWN_EXTENSION ...]
                        Extensions from markdown module, defaults to ['codehilite', 'tables', "toc(marker='')", 'meta']
  -O OUTPUT_FILE, --output_file OUTPUT_FILE
                        HTML output file, defaults to stdout

```

- **Additional CSS styles can be defined in the markdown file metadata**
    - It is assumed they're located in `/css` subfolder
- **Base HTML template** can be set, defaults to `template.html`
- **Logo, favicon and background images can be set directly in the markdown**
- Any setting passed via CLI overrides these in the markdown, except for CSS (all are applied)

## Recognized metadata in markdown file:

- **date**: Document's creation date. If omitted, current date is shown instead

## Recognized by default template file:

- **Title**: Document's title/s
- **css**: name of the CSS files under `/css` subfolder
- **template**: name of the base template under `/layout` subfolder
- **Author**: Author/s of the document
- **email**: email addresses for each author
- **logo**: file or URL for the optional logo image
- **logolink**: valid URL for the logo
- **theme**: anything here will be passed as: `<body class="{{ }}">`
- **background_img**: background image
- **favicon**: favicon file (not .ico)

## [TODO]

* ~~Let CSS and templates being in any location~~
* ~~Allow setting a list of templates directly in the markdown file~~ moved most part of the logic to the templates
* Allow both HTML-embedded and online HTML (do not embed CSS or images for online resources)
* Fix TOC to allow pandoc's-like `--toc-depth`
* Conversion to other formats
* ~~Global speedup, refactoring, ...~~
* ~~Check if logolink contains a valid URL~~ moved extra logic to templates
* Complete refactoring


changelog.md
- v0.0.4 - Fix tests
- v0.0.3 - Remove many layout-specific settings from input args, small refactor
- v0.0.2 - Working version with settings either in markdown or input arguments
- v0.0.1 - Initial version

test/tests.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""
md2html - test functions
"""

import unittest
from md2html import md2html

from cStringIO import StringIO

import os
import tempfile
from contextlib import contextmanager

@contextmanager
def tempinput(data):
    temp = tempfile.NamedTemporaryFile(delete=False)
    temp.write(data)
    temp.close()
    yield temp.name
    os.unlink(temp.name)

TESTS_DIR =os.path.dirname(__file__)

class TestMd2html(unittest.TestCase):
    """ Set of test functions for md2html module """

    def __init__(self, *args, **kwargs):
        super(TestMd2html, self).__init__(*args, **kwargs)
        self.test = md2html.MD2Html()
        self.test.logger.setLevel('DEBUG')

    def test_to_b64(self):
        """ Test function for MD2Html.to_b64() """
        local_image = "md2html/layout/images/Acme-logo.png"
        remote_image = "http://upload.wikimedia.org/wikipedia/"\
                       "commons/7/7e/Acme-logo.png"
        self.assertIsNot(self.test.to_b64(local_image), '')
        self.assertIsNot(self.test.to_b64(remote_image), '')
        self.assertEqual(self.test.to_b64(local_image),
                         self.test.to_b64(remote_image))
        self.assertIs(self.test.to_b64('nonexisting_file.png'), '')
        self.assertIs(self.test.to_b64('http://localhost/nofile.jpg'), '')
        self.assertIs(self.test.to_b64('http://wrong.url/img.png'), '')

    def test_reencodehtml(self):
        """ Test function for MD2Html.reencode_html() """
        html = ['<html><body><h1>TITLE</h1><a>Hello World!</a>',
                '</body></html>']
        self.assertEqual(self.test.reencode_html(html),
                         ''.join(html))
        html = ['<html><body><h1>TITLE</h1><a>Hello World!</a>',
                '<img src="md2html/layout/images/Acme-logo.png" />',
                '</body></html>']
        self.assertIn('data:image/png;base64,', self.test.reencode_html(html))

    def test_csscompress(self):
        """ Test function for MD2Html.css_compress() """
        self.assertListEqual([k for k in self.test.css_compress([])], [])
        css_files = ['md2html/css/lanyon.css',
                     'md2html/css/codehilite.css']
        css_comp1 = self.test.css_compress(css_files)
        self.assertIsNotNone(''.join([css for css in css_comp1]))

        css_comp1 = self.test.css_compress(css_files)
        css_comp2 = self.test.css_compress(css_files + ['nonexisting.css'])
        self.assertEqual(''.join([css for css in css_comp1]),
                         ''.join([css for css in css_comp2]))

    def test_gethtml(self):
        """ Test function for MD2Html.get_html() """
        md_string = '#Title\n##This is my markdown file!\n![Alt](!'\
                    'https://duckduckgo.com/assets/badges/logo_square.64.png'\
                    ')\n`End of file`'
        with tempinput(md_string) as temp_md:
            self.test.md_extensions = []
            html = self.test.get_html(temp_md)
            self.assertMultiLineEqual(html[0],
                                      '<h1>Title</h1>\n'
                                      '<h2>This is my markdown file!</h2>\n'
                                      '<p><img alt="Alt" src="!https://duckduckgo'
                                      '.com/assets/badges/logo_square.64.png" />\n'
                                      '<code>End of file</code></p>')
            self.assertIsNone(html[1])
            self.assertIsNone(html[2])

            self.test.md_extensions = md2html.MD_EXTENSIONS
            html = self.test.get_html(temp_md)
            self.assertDictEqual(html[1], {})  # empty metadata

        with tempinput('metadata: metadata example\n%s' % md_string) as temp_md:
            html = self.test.get_html(temp_md)
            map(self.assertIsNotNone, html)
            self.assertDictEqual(html[1],
                                 {u'metadata': [u'metadata example']})

    def test_main(self):
        """ End to end test of overall module by running MD2Html.main()
        """
        tmp_file = StringIO()

        self.test.main(md_file='test/test.md',
                       working_dir='md2html',
                       output_file=tmp_file)
        with open('%s/test.html' % TESTS_DIR, 'r') as html:
            that = html.read()
        tmp_file.seek(0)
        this = tmp_file.read()
        # self.assertMultiLineEqual(this, that)
        self.assertRaises(OSError, self.test.main, md_file='not_a_test.md')


if __name__ == '__main__':
    unittest.main()


test/test_md2html.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""
md2html - test functions
"""

import unittest
import md2html
import filecmp
from cStringIO import StringIO


class TestMd2html(unittest.TestCase):
    """ Set of test functions for md2html module """

    def __init__(self, *args, **kwargs):
        super(TestMd2html, self).__init__(*args, **kwargs)
        self.test = md2html.MD2Html()
        self.test.logger.setLevel('DEBUG')

    def test_to_b64(self):
        """ Test function for MD2Html.to_b64() """
        local_image = "../md2html/layout/images/Acme-logo.png"
        remote_image = "http://upload.wikimedia.org/wikipedia/"\
                       "commons/7/7e/Acme-logo.png"
        self.assertIsNotNone(self.test.to_b64(local_image))
        self.assertIsNotNone(self.test.to_b64(remote_image))
        self.assertEqual(self.test.to_b64(local_image),
                         self.test.to_b64(remote_image))
        self.assertIsNone(self.test.to_b64('nonexisting_file.png'))
        self.assertIsNone(self.test.to_b64('http://localhost/nofile.jpg'))
        self.assertIsNone(self.test.to_b64('http://wrong.url/img.png'))

    def test_reencodehtml(self):
        """ Test function for MD2Html.reencode_html() """
        html = ['<html><body><h1>TITLE</h1><a>Hello World!</a>',
                '</body></html>']
        self.assertEqual(self.test.reencode_html(html),
                         ''.join(html))
        html = ['<html><body><h1>TITLE</h1><a>Hello World!</a>',
                '<img src="../md2html/layout/images/Acme-logo.png" />',
                '</body></html>']
        self.assertIn('data:image/png;base64,', self.test.reencode_html(html))

    def test_csscompress(self):
        """ Test function for MD2Html.css_compress() """
        self.assertListEqual([k for k in self.test.css_compress([])], [])
        css_files = ['../md2html/css/lanyon.css',
                     '../md2html/css/codehilite.css']
        css_comp1 = self.test.css_compress(css_files)
        self.assertIsNotNone(''.join([css for css in css_comp1]))

        css_comp1 = self.test.css_compress(css_files)
        css_comp2 = self.test.css_compress(css_files + ['nonexisting.css'])
        self.assertEqual(''.join([css for css in css_comp1]),
                         ''.join([css for css in css_comp2]))

    def test_gethtml(self):
        """ Test function for MD2Html.get_html() """
        md_string = '#Title\n##This is my markdown file!\n![Alt](!'\
                    'https://duckduckgo.com/assets/badges/logo_square.64.png'\
                    ')\n`End of file`'

        fbuffer = StringIO()
        fbuffer.write(md_string)

        fbuffer.seek(0)
        self.test.md_extensions = []
        html = self.test.get_html(fbuffer)
        self.assertMultiLineEqual(html[0],
                                  '<h1>Title</h1>\n'
                                  '<h2>This is my markdown file!</h2>\n'
                                  '<p><img alt="Alt" src="!https://duckduckgo'
                                  '.com/assets/badges/logo_square.64.png" />\n'
                                  '<code>End of file</code></p>')
        self.assertIsNone(html[1])
        self.assertIsNone(html[2])

        fbuffer.seek(0)
        self.test.md_extensions = md2html.MD_EXTENSIONS
        html = self.test.get_html(fbuffer)
        self.assertDictEqual(html[1], {})  # empty metadata
        fbuffer.close()

        fbuffer = StringIO()
        fbuffer.write('metadata: metadata example\n%s' % md_string)
        fbuffer.seek(0)
        html = self.test.get_html(fbuffer)
        map(self.assertIsNotNone, html)
        self.assertDictEqual(html[1],
                             {u'metadata': [u'metadata example']})
        fbuffer.close()

    def test_main(self):
        """ End to end test of overall module by running MD2Html.main()
        """
        tmp_file = StringIO()
        with open('test.md', 'r') as md_fileobject:
            self.test.main(md_fileobject=md_fileobject,
                           output_file=tmp_file,
                           working_dir='../md2html')
        with open('test.html', 'r') as htmlfile:
            that = htmlfile.read()
        tmp_file.seek(0)
        this = tmp_file.read()
        self.assertMultiLineEqual(this, that)


if __name__ == '__main__':
    unittest.main()


mdtohtml/md2html.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""
Created on Tue Apr  7 17:18:42 2015

@author: fernandezjm
"""

import jinja2
import markdown
import os
import re
import imghdr
import logging
import codecs

from cStringIO import StringIO
from rfc3987 import parse
from urllib2 import urlopen, HTTPError, URLError
from csscompressor import compress
from datetime import date
from dateutil.parser import parse as date_parse

HTML_TEMPLATE = 'template.html'
MD_EXTENSIONS = ["codehilite", "tables", "toc(marker='')", "meta"]
DEFAULT_LOGLEVEL = 'INFO'


class MyLoader(jinja2.BaseLoader):

    def __init__(self, sourcepath, logger):
        self.path = os.path.dirname(sourcepath)
        self.logger = logger

#TODO: write a test function for including raw HTML
    def get_source(self, environment, template):
        path = os.path.abspath(os.path.relpath(template, self.path))
        self.logger.debug('Rendering %s', path)

        if not os.path.exists(path):
            raise jinja2.TemplateNotFound(template)
        mtime = os.path.getmtime(path)
        with file(path) as f:
            source = f.read().decode('utf-8')
        return source, path, lambda: mtime == os.path.getmtime(path)


def init_logger(loglevel=None, name=__name__):
    """ Initialize logger, sets the appropriate level and attaches a console
        handler.
    """
    logger = logging.getLogger(name)
    logger.setLevel(loglevel or DEFAULT_LOGLEVEL)

    # If no console handlers yet, add a new one
    if not any(isinstance(x, logging.StreamHandler) for x in logger.handlers):
        console_handler = logging.StreamHandler()
        if logging.getLevelName(logger.level) == 'DEBUG':
            _fmt = '%(asctime)s| %(levelname)-8s| %(threadName)10s/' \
                   '%(lineno)03d@%(module)-10s| %(message)s'
            console_handler.setFormatter(logging.Formatter(_fmt))
        else:
            console_handler.setFormatter(
                logging.Formatter('%(asctime)s| %(levelname)-8s| %(message)s'))
        logger.addHandler(console_handler)

    logger.info('Initialized logger with level: %s',
                logging.getLevelName(logger.level))
    return logger


class MD2Html(object):

    """
    Main class object
    """

    def __init__(self):
        self.working_dir = os.path.dirname(os.path.abspath(__file__))
        self.html_template = HTML_TEMPLATE
        self.md_extensions = MD_EXTENSIONS
        self.logger = init_logger()

    def main(self, md_file, **kwargs):
        """ Renders an html from jinja2 templates filled with content converted
            from markdown source
            Mandatory arguments:
            - md_contents: markdown input file content (passed to get_html)

            Optional arguments:

            - html_template:  Template for jinja2
            - output_file:    Writtable output file descriptor object
            - md_extensions:  List of extensions for markdown module
            - working_dir:    Directory where /layout and /css are found
            - loglevel:       Debug level passed to logging
        """
        # Remove all "None" input values
        list(map(kwargs.pop, [item for item in kwargs if not kwargs[item]]))
        if 'loglevel' in kwargs:
            self.logger.setLevel(kwargs.get('loglevel'))
        self.html_template = kwargs.get('html_template', self.html_template)
        self.working_dir = kwargs.get('working_dir', self.working_dir)

        if not os.path.isfile('%s/layout/%s' % (self.working_dir,
                                                self.html_template)):
            self.logger.error('Base template "%s" not found in %s/layout.',
                              self.html_template,
                              self.working_dir)
            return

        self.md_extensions = kwargs.get('md_extensions', self.md_extensions)
        output_file_descr = kwargs.get('output_file', None)

        self.logger.debug('Loading template %s/layout/%s',
                          self.working_dir,
                          self.html_template)
        loader = jinja2.FileSystemLoader('{}/layout'.format(self.working_dir))
        environment = jinja2.Environment(loader=loader, trim_blocks=True)

#       FUNCTIONS PASSED TO JINJA2
        environment.globals['include_raw'] = \
            lambda filename: jinja2.Markup(loader.get_source(environment,
                                                             filename)[0]) \
            if os.path.exists(filename) else ''

        environment.globals['hrefs_from_keypair'] = \
            lambda keypair: [u'<a href={1}>{0}</a>'.format(*item.split(','))
                             for item in keypair]

        environment.globals['hrefs_from_lists'] = \
            lambda names, links: [''] if len(names) != len(links) else \
            [u'<a href={}>{}</a>'.format(links[i], name) for (i, name)
             in enumerate(names)]

        environment.globals['css_compress'] = self.css_compress

        environment.globals['to_base64'] = \
            lambda image: self.to_b64(image[0])
#       END OF FUNCTIONS PASSED TO JINJA2

        template = environment.get_template(self.html_template)
        content, metadata, toc = self.get_html(md_file)

        try:
            markdown_date = date_parse(metadata['date'][0]) \
                            if 'date' in metadata else date.today()
        except ValueError:  # date in markdown not recognized by parser
            print "Date in markdown not recognized by parser (wrong locale?)"
            markdown_date = None

        fbuffer = StringIO()
        rendered = template.render(content=content,
                                   metadata=metadata,
                                   toc=toc,
                                   date=markdown_date,
                                   **kwargs)

        # MD may contain a {% include 'that.html' %}, 2nd rendering pass needed
        if kwargs.get('2pass', None):
            self.logger.info('Second pass rendering MD Jinja2 references')
            loader = MyLoader(os.path.abspath(md_file), self.logger)
            environment = jinja2.Environment(loader=loader, trim_blocks=True)
            template = environment.from_string(rendered)
            fbuffer.write(template.render().encode('utf-8'))
        else:
            fbuffer.write(rendered.encode('utf-8'))

        fbuffer.seek(0)
        self.logger.debug('Rendering OK')

        if output_file_descr:
            output_file_descr.write(self.reencode_html(fbuffer))
        else:
            print(self.reencode_html(fbuffer))
        fbuffer.close()
        return

    def to_b64(self, image_filename, *args):
        """ Returns a tuple with (b64content, imgtype) where:
            - b64content is a base64 representation of the input file
            - imgtype is the image type as detected by imghdr
        """
        self.logger.debug('Converting image %s to base64', image_filename)
        self.logger.debug('Current directory %s', os.path.abspath(os.curdir))
        try:
            img_info = parse(image_filename, rule='IRI')
            extension = img_info['path'].split('.')[-1]
            content = urlopen(image_filename)
        except ValueError:  # not a valid IRI, assume local file
            self.logger.debug("Image '%s' doesn't have a valid URL, "
                              "assuming local", image_filename)
            try:
                extension = imghdr.what(image_filename)
                if extension is None:
                    self.logger.debug('Image extension not detected, skipping')
                    return ''
                content = open(image_filename, 'rb')
            except (IOError, AttributeError, TypeError):
                return ''
        except (HTTPError, URLError, TypeError):
            return ''
        txt = 'data:image/{};base64,\n{}'.format(extension,
                                                 content.read().encode('base64'
                                                                       )
                                                 )
        content.close()
        return txt

    def reencode_html(self, input_html):
        """
        Embeds all images from an html file to base64
        """
        fbuffer = StringIO()
        my_regex = re.compile('.*(<img.+src=\")(.+?)"', re.IGNORECASE)
        for line in input_html:
            try:
                item = re.finditer(my_regex, line).next().group(2)
                self.logger.debug('Embedding image: %s',
                                  item)
                fbuffer.write(line.replace(item, self.to_b64(item)))
            except StopIteration:
                fbuffer.write(line)
        fbuffer.seek(0)
        content = fbuffer.read()
        fbuffer.close()
        return content

    def css_compress(self, css_files):
        """ Reads file by file and returns the compressed version of the CSS
        css_files: list of files, located in /css subfolder
        """
        def css_wrapper(css_files, ):
            """ Wrapper for css_compress handling with non-existing files """
            for item in css_files:
                try:
                    self.logger.debug('Loading CSS %s/css/%s',
                                      self.working_dir,
                                      item)
                    with open('{}/css/{}'.format(self.working_dir, item),
                              'r') as css_file:
                        yield '/* CSS from metadata: {} */ {}'.format(
                                                     item,
                                                     compress(css_file.read()))
                except IOError:
                    yield ''

        return ''.join([css for css in css_wrapper(css_files)])

    def get_html(self, md_file):
        """ Converts markdown syntax to html
             - md_file: markdown input file name

            Returns:
            - html:    HTML encoded string
            - md.Meta: Metadata as found in md header (if any)
            - md.toc:  Table of Contents, if TOC in md_extensions
        """
        extensions = ['markdown.extensions.%s' % k for k in self.md_extensions]
        self.logger.debug('Generating markdown with extensions: %s',
                          ', '.join(extensions))
        md_converter = markdown.Markdown(extensions=extensions)
        md_content = self.read_mdfile(md_file)
        html = md_converter.convert(md_content)

        if any(['toc' in k for k in self.md_extensions]):
            if 'meta' in self.md_extensions:
                return (html, md_converter.Meta, md_converter.toc)
            else:
                return (html, None, md_converter.toc)
        else:
            if 'meta' in self.md_extensions:
                return (html, md_converter.Meta, None)
            else:
                return (html, None, None)

    def read_mdfile(self, input_file):
        """ Open the input file detecting the encoding (UTF-8 with BOM) """
        try:
            bytes = min(32, os.path.getsize(input_file))
            with open(input_file, 'rb') as raw:
                header = raw.read(bytes)

                if header.startswith(codecs.BOM_UTF8):
                    encoding = 'utf-8-sig'
                elif header.startswith(codecs.BOM_UTF16_LE) or \
                     header.startswith(codecs.BOM_UTF16_BE):
                    encoding = 'utf-16'
                elif header.startswith(codecs.BOM_UTF32_LE) or \
                     header.startswith(codecs.BOM_UTF32_BE):
                    encoding = 'utf-32'
                else:
                    encoding = 'utf-8'

                raw.seek(0)
                data = raw.read().decode(encoding)
            return data
        except OSError as exception:
            self.logger.error('No such file: %s', input_file)
            raise exception


mdtohtml/__init__.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""
Created on Tue Apr  7 17:18:42 2015

@author: fernandezjm
"""

import os, sys
import argparse
import md2html


def argument_parse():
    """ Argument parser for main method
    """
    MD_TO_HTML = md2html.MD2Html()
    parser = argparse.ArgumentParser(formatter_class=argparse.
                                     RawTextHelpFormatter,
                                     description='Markdown to HTML converter')

    parser.add_argument('-C', '--css', metavar='CSS_FILE',
                        dest='css_files', nargs='+', default=[],
                        type=str, help='CSS files')

    parser.add_argument('-W', '--working-dir', metavar='BASE_FOLDER',
                        dest='working_dir', default=MD_TO_HTML.working_dir,
                        type=str, help='Working base directory where '
                                       '/css and /layout folders are found')

    parser.add_argument('-T', '--template',
                        dest='html_template', default=MD_TO_HTML.html_template,
                        type=str, help='HTML template for Jinja2')

    parser.add_argument('-M', '--extensions', metavar='MARKDOWN_EXTENSION',
                        dest='md_extensions', nargs='+', type=str,
                        default=MD_TO_HTML.md_extensions,
                        help='Extensions from markdown module, defaults to '
                             '{}'.format(MD_TO_HTML.md_extensions))

    parser.add_argument('md_file',
                        type=str, metavar='input_file',
                        help='Markdown input file')

    parser.add_argument('-O', '--output_file', default=sys.stdout,
                        type=argparse.FileType('w'),
                        help='HTML output file, defaults to stdout')

    parser.add_argument('-2', action='store_true', dest='2pass',
                        help='Make a 2pass render, in case markdown file '
                             'contains Jinja2 syntax (i.e. {%% includes %%})')

    parser.add_argument('--loglevel', const="WARNING",
                        choices=['DEBUG',
                                 'INFO',
                                 'WARNING',
                                 'ERROR',
                                 'CRITICAL'],
                        help='Debug level (default: {})'.format(
                            md2html.DEFAULT_LOGLEVEL),
                        nargs='?')

    userargs = vars(parser.parse_args())

    if len(sys.argv) == 1:
        parser.print_help()
    else:
        MD_TO_HTML.main(**userargs)


if __name__ == "__main__":
    argument_parse()
